dcos/nodeutil/util.go

```diff
@@ -0,0 +1,391 @@
+package nodeutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/dcos/dcos-go/dcos"
+	"github.com/dcos/dcos-go/exec"
+)
+
+const (
+	defaultExecTimeout       = 10
+	defaultClusterIDLocation = "/var/lib/dcos/cluster-id"
+	defaultBashPath          = "/bin/bash"
+)
+
+// ErrNodeInfo is an error structure raised by exported functions with meaningful error message.
+type ErrNodeInfo struct {
+	msg string
+}
+
+func (e ErrNodeInfo) Error() string {
+	return e.msg
+}
+
+// NodeInfo defines an interface to interact with DC/OS cluster via go methods.
+type NodeInfo interface {
+	DetectIP() (net.IP, error)
+	Role() (string, error)
+	IsLeader() (bool, error)
+	MesosID(context.Context) (string, error)
+	ClusterID() (string, error)
```

Comment:
One thing to consider is possibly having a AgentNode{} and a MasterNode{} since things like `ClusterID` are not available on agents, and therefore that method to get it would bail out in a generic interface like this.